### PR TITLE
SetGridField automation

### PR DESCRIPTION
**Pull request by GalRabin**
#6335  Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/19484

## Description
Automation for populating grid in incident layout.
### Design
**Args:**
 1. `context_paths` -Context paths comma separated (same number of availble columns in grid) e.g. a.b.c,a.d.e (with 2 columns).
 2. `sort_by` -  Column name to sort by - e.g col1.
 3.  `grid_id` - Could be gathered from Context Data -> Incident (I normalize it case in camelcase or cases).
  4. `row_correlation` - true if row represents an object, false otherwise.
  5. `overwrite` - true if remove old-data, false to append data.
  6. `columns` - Columns names as shown in the grid.

**Limitations:**
* The columns of the grid can't be modified - name and the amount of the columns (server limitations, verified it with @Bars92 .
* Can't validate that the user gives correct input, due to an unknown number of columns when using API call to demisto, not consistent behavior, so the user must know what he is doing, otherwise, he will not see any effect.
for example: 
    1. When the grid is empty I will get no columns (so I can't know how column there are or what their name is.
    2. When the gird is not empty field, I will get the column that only field (so I can get 1 out of 2, so I can't give error on that).

## Screenshots
Example for usage:
```
!SetGridField context_paths="EWS.Items.headers.name,EWS.Items.headers.value,EWS.Items.headers.ourValue" grid_id='my_grid' overwrite=false columns="heade1,header2,heaad3"
```
Example for populating grid:
image

## Minimum version of Demisto
- [ ] 4.5.0
- [ ] 5.0.0
- [x] 5.5.0

## Does it break backward compatibility?
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by idovandijk** [comment]
Please note my comment in the issue before merging

File: Packs/Legacy/Scripts/SetGridField/SetGridField.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+                        overwrite=overwrite,
+                        current_table=current_table)
+
+    return table
+
+
+def main():
+    try:
+        # Normalize grid id from any form to connected lower words, e.g. my_word/myWord -> myword
+        grid_id = phrases_case.camel(demisto.getArg('grid_id')).replace("'", "").lower()
```
[the remark this answers is not in the document]
Comment (GalRabin) [reply Packs/Legacy/Scripts/SetGridField/SetGridField.py:233]: from some reason it return as repr and not as str, so I need to remove them.

File: Packs/Legacy/Scripts/SetGridField/SetGridField.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+def build_grid_command(grid_id: str, context_paths: str, sort_by: Optional[str], row_correlation: bool, overwrite: bool,
+                       columns: Optional[str]) -> List[Dict[str, str]]:
+    """ Build Table as dictionary for populate grid in incident layout.
+
+    The required table structure to populate in the incident is as follow:
+    [{'col1': 'val1'},{'col2': 'val2'},{'col3': 'val3'},{'col4': 'val4'}] which represent:
+    | col1 | col2 |
+    | ---- | ---- |
+    | val1 | val2 |
+    | val3 | val4 |
```
[the remark this answers is not in the document]
Comment (GalRabin) [reply Packs/Legacy/Scripts/SetGridField/SetGridField.py:199]: Forgot to update description, fixed

File: Packs/Legacy/Scripts/SetGridField/SetGridField.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+              {'col4': 'val4'}].
+        list: Table columns name.
+     """
+    # Get current Grid data
+    current_table: Optional[List[dict]] = demisto.incidents()[0].get("CustomFields", {}).get(grid_id)
+    if not current_table:
+        raise ValueError(f"The grid id isn't valid : {grid_id}")
+    # Validate columns number the same as context paths - If no data initiated skip validation, but check if columns specified
+    if not columns:
+        raise ValueError("Columns not specified - Its a mandatory arg when grid is empty.")
```
[the remark this answers is not in the document]
Comment (GalRabin) [reply Packs/Legacy/Scripts/SetGridField/SetGridField.py:45]: Fixed, I can't know, Only validate that context_path and column are equal

File: Packs/Legacy/Scripts/SetGridField/SetGridField.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+                current_absolute_path = '..'.join(splited_path[:-1])
+                if absolute_path:
+                    if absolute_path != current_absolute_path:
+                        raise ValueError(
+                            f'Unable to correlate rows due to keys not in the same context path {current_absolute_path}')
+                else:
+                    absolute_path = current_absolute_path
+                correlated_keys.append(splited_path[-1])
+            else:
+                raise ValueError(f'Unable to correlate rows due to not valid context path {context_path}')
```
[the remark this answers is not in the document]
Comment (GalRabin) [reply Packs/Legacy/Scripts/SetGridField/SetGridField.py:137]: I need to populate the not full dict to be full in order to use pandas, So I gather  the keys to populate if missing

File: Packs/Legacy/Scripts/SetGridField/SetGridField.yml [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+enabled: false
+name: SetGridField
+script: '-'
+subtype: python3
+system: false
+tags:
+- ''
+timeout: '0'
+type: python
+dockerimage: devdemisto/grid:1.0.0.7488
```
[the remark this answers is not in the document]
Comment (GalRabin) [reply Packs/Legacy/Scripts/SetGridField/SetGridField.yml:62]: Sure

**Comment by GalRabin** [comment]
@kirbles19
I will appreciate your review on the following:
1. [Readme](https://github.com/demisto/content/blob/SetGridField/Packs/Legacy/Scripts/SetGridField/README.md) - all lines
2. [Yaml](https://github.com/demisto/content/blob/SetGridField/Packs/Legacy/Scripts/SetGridField/SetGridField.yml) - all files

File: Packs/Legacy/Scripts/SetGridField/CHANGELOG.md [hunk 6]
```diff
@@ -0,0 +1,2 @@
+## [Unreleased]
```
Comment (yaakovi) [inline Packs/Legacy/Scripts/SetGridField/CHANGELOG.md:1]: remove this file

File: Packs/Legacy/Scripts/SetGridField/README.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+![Grid](grid.png)
+
+Considering the following cases:
+1. Key value to Grid:
+```shell script
+!SetGridField columns="columnheade1,columnheade2,columnheade3" context_path=EWS.Items.HeadersMap grid_id=mygrid keys="Received,Thread-Index"
+```
+
+Grid after update: \
+![Grid](grid_key_value_update.png)
```
Comment (yaakovi) [inline Packs/Legacy/Scripts/SetGridField/README.md:69]: update image links according to https://xsoar.pan.dev/docs/integrations/integration-docs#images

File: Packs/Legacy/Scripts/SetGridField/README.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+          "value": "value4"
+        }
+      ]
+    }
+  }
+}
+```
+
+2. Grid: \
+![Grid](grid.png)
```
Comment (yaakovi) [inline Packs/Legacy/Scripts/SetGridField/README.md:60]: update image links according to https://xsoar.pan.dev/docs/integrations/integration-docs#images

File: Packs/Legacy/Scripts/SetGridField/README.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+Grid after update: \
+![Grid](grid_key_value_update.png)
+ 
+2. List of item properties to Grid:
+```shell script
+!SetGridField columns="columnheade1,columnheade2,columnheade3" context_path=EWS.Items.headers grid_id=mygrid keys="Received,Thread-Index"
+```
+
+Grid after update: \
+![Grid](grid_list_update.png)
```
Comment (yaakovi) [inline Packs/Legacy/Scripts/SetGridField/README.md:77]: update image links according to https://xsoar.pan.dev/docs/integrations/integration-docs#images

File: Packs/Legacy/Scripts/SetGridField/SetGridField_test/context_entry_list.json [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    "value": "value1"
+  },
+  {
+    "name": "name1",
+    "value": "value1"
+  },
+  {
+    "name": "name1",
+    "value": "value1"
+  }
```
Comment (yaakovi) [inline Packs/Legacy/Scripts/SetGridField/SetGridField_test/context_entry_list.json:17]: please update values with ascending numbers (1, 2, 3, 4)

File: Packs/Legacy/Scripts/SetGridField/SetGridField.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (yaakovi) [reply Packs/Legacy/Scripts/SetGridField/SetGridField.yml:58]: reminder

File: Packs/Legacy/Scripts/SetGridField/SetGridField.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+
+def main():
+    try:
+        # Normalize grid id from any form to connected lower words, e.g. my_word/myWord -> myword
+        grid_id = normalized_string(demisto.getArg('grid_id'))
+        # Build updated table
+        table = build_grid_command(grid_id=grid_id,
+                                   context_path=demisto.getArg('context_path'),
+                                   keys=argToList(demisto.getArg('keys')),
+                                   overwrite=False if demisto.getArg('overwrite') == 'false' else True,
```
Comment (yaakovi) [inline Packs/Legacy/Scripts/SetGridField/SetGridField.py:184]: invalid input will not override
```suggestion
                                   overwrite=demisto.getArg('overwrite').lower() == 'true',
```

File: Packs/Legacy/Scripts/SetGridField/SetGridField.py [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+
+def build_grid_command(grid_id: str, context_path: str, keys: List[str], columns: List[str], overwrite: bool, sort_by: str) \
+        -> List[Dict[Any, Any]]:
+    """ Build Grid in one of the 2 options:
+            1. Context_path contain list, e.g. [{'a': 1, 'b': 2}, {'a': 1, 'b': 2}]
+            2. Context_path contain dict (key value pairs), e.g. {'a': 1, 'b': 2}
+
+        Warnings:
+            1. The automation can't validate that the columns name correct and its mandatory.
+            2. The automation knows how to handle only list or dict and doesn't know hoe to handle.
```
Comment (yaakovi) [inline Packs/Legacy/Scripts/SetGridField/SetGridField.py:145]: fix line

File: Packs/Legacy/Scripts/SetGridField/SetGridField.py [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+    # Validate entry context structure
+    validate_entry_context(entry_context_data, keys)
+    # Building new Grid
+    if len(entry_context_data) > 1:
+        # Handle entry context list option
+        entry_context_data = [filter_the_dict(item, keys, len(columns)) for item in entry_context_data]
+        table = pd.DataFrame(entry_context_data)
+        table.rename(columns=dict(zip(table.columns, columns)), inplace=True)
+    elif len(entry_context_data) == 1 and isinstance(entry_context_data[0], dict):
+        # Handle entry context key, vlaue option
```
Comment (yaakovi) [inline Packs/Legacy/Scripts/SetGridField/SetGridField.py:128]:
```suggestion
        # Handle entry context key-value option
```

File: Packs/Legacy/Scripts/SetGridField/SetGridField.py [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+        for item in entry_context:
+            if not isinstance(item, dict):
+                raise ValueError(exception_msg)
+            else:
+                for key, value in item.items():
+                    if key in keys:
+                        if not (isinstance(key, str) and isinstance(value, str)):
+                            raise ValueError(exception_msg)
+    else:
+        raise ValueError(exception_msg)
```
Comment (yaakovi) [inline Packs/Legacy/Scripts/SetGridField/SetGridField.py:98]: minimize indentations
```suggestion
    if not isinstance(entry_context, list):
        raise ValueError(exception_msg)

    for item in entry_context:
        if not isinstance(item, dict):
            raise ValueError(exception_msg)
            
        for key, value in item.items():
            if key in keys:
                if not (isinstance(key, str) and isinstance(value, str)):
                    raise ValueError(exception_msg)
```

File: Packs/Legacy/Scripts/SetGridField/SetGridField.py [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+    """
+    exception_msg = "Not valid entry context path - dict[str,str] or list[dict[str,str]]"
+    if isinstance(entry_context, list):
+        for item in entry_context:
+            if not isinstance(item, dict):
+                raise ValueError(exception_msg)
+            else:
+                for key, value in item.items():
+                    if key in keys:
+                        if not (isinstance(key, str) and isinstance(value, str)):
```
Comment (yaakovi) [inline Packs/Legacy/Scripts/SetGridField/SetGridField.py:95]: why value must be a string? what if it is a number for example?

File: Packs/Legacy/Scripts/SetGridField/SetGridField.py [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+        "testword"
+        >>> normalized_string("hello_world")
+        "hello_world"
+    """
+    return phrases_case.camel(phrase).replace("'", "").lower()
+
+
+def filter_the_dict(dict_obj: Dict[Any, Any], keys: List[str], max_keys: Optional[int] = None) -> Dict[Any, Any]:
+    """ Filter keys from Dictionary:
+            1. Will save all keys in keys parameter.
```
Comment (yaakovi) [inline Packs/Legacy/Scripts/SetGridField/SetGridField.py:32]: `keys` does not change in the code (and it shouldn't be)

File: Packs/Legacy/Scripts/SetGridField/SetGridField.py [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+    Examples:
+        >>> normalized_string("TestWord")
+        "testword"
+        >>> normalized_string("hello_world")
+        "hello_world"
+    """
+    return phrases_case.camel(phrase).replace("'", "").lower()
+
+
+def filter_the_dict(dict_obj: Dict[Any, Any], keys: List[str], max_keys: Optional[int] = None) -> Dict[Any, Any]:
```
Comment (yaakovi) [inline Packs/Legacy/Scripts/SetGridField/SetGridField.py:30]:
```suggestion
def filter_dict(dict_obj: Dict[Any, Any], keys: List[str], max_keys: Optional[int] = None) -> Dict[Any, Any]:
```

File: Packs/Legacy/Scripts/SetGridField/SetGridField.py (hunk 15 above)
Comment (GalRabin) [reply Packs/Legacy/Scripts/SetGridField/SetGridField.py:95]: Applied int float, int, None also

File: Packs/Legacy/Scripts/SetGridField/SetGridField.py (hunk 16 above)
Comment (GalRabin) [reply Packs/Legacy/Scripts/SetGridField/SetGridField.py:32]: I wrote more clear description

**Comment by kirbles19** [comment]
@GalRabin Done. Please run the doc script on the edited YML. Too many changes to do manually on the README

**Pull request merged**